### ISSUE TITLE
ENA-115: Overview: Redesign save alerts to provide more info when project can’t save (Phase I)

### ISSUE THREAD

**Issue by kaaaaathy** [LLK/scratch-gui#6069]
# Links
### Specs: Customized alerts
- **[An asset is too big (most common and most problematic)](https://github.com/LLK/scratch-gui/issues/6073)**
- **[User gets signed out](https://github.com/LLK/scratch-gui/issues/6074)**
- **[Connectivity, can be Scratch side, or user side](https://github.com/LLK/scratch-gui/issues/6075)**
### Specs: Changes to nav
- **[Nav: Save status](https://github.com/LLK/scratch-gui/issues/6080)**
- **[Nav: Connectivity status](https://github.com/LLK/scratch-gui/issues/6081)**

### General spec
- **[Invision design file](https://invis.io/N7XWQU2B5K3)**

----------------------------

# Background
### Currently...
[This design](https://github.com/LLK/scratch-gui/issues/3761#issuecomment-442544575) I made is what is currently implemented.
Here is an overview of the issues and pain points that we are solving in Phase I.

![OVERVIEW: As Is](https://user-images.githubusercontent.com/8203939/89483873-1dd21080-d76b-11ea-89fe-dc1a7f402258.png)

(There is a 4th pain point not listed in this image, which is that you can't create or upload a different project because you get stuck on this error.)

### Related filed issues
https://github.com/LLK/scratch-gui/issues/4045
https://github.com/LLK/scratch-gui/issues/4225
https://github.com/LLK/scratch-gui/issues/3764

# Overview of Proposed Changes
### Goals

1. **Customize alerts to each situation (more info below)**
— User can know how to debug or seek help
2. **Show save status in nav**
 — User can continue working offline if they want, without alert as distraction
3. **Show connectivity more clearly** 
— User knows they're offline, separate from save status

![image](https://user-images.githubusercontent.com/8203939/89483985-68538d00-d76b-11ea-8de0-43120a7a57a6.png)

# Goal 1: Customize alerts to each situation

- ### [An asset is too big (most common and most problematic)](https://github.com/LLK/scratch-gui/issues/6073)
- ### [User gets signed out](https://github.com/LLK/scratch-gui/issues/6074)
- ### [Connectivity, can be Scratch side, or user side](https://github.com/LLK/scratch-gui/issues/6075)
- ### [Generic alert for any other case](https://github.com/LLK/scratch-gui/issues/6076)

----------------
### What about multiple alerts?
There will generally be no situations where there are **more than 1 of EACH TYPE of alert.**
This is the order of operations in which they should appear:

![7 1 3](https://user-images.githubusercontent.com/8203939/89484311-3c84d700-d76c-11ea-88bc-e7a2863a15fb.png)

### HOWEVER...
There is a case within Use Case 1 where the user may encounter multiple "Asset too big" issues. On average it's between 1-3, but we've seen cases of ~5 or more on occasion.

In this case, follow this defined behavior for multiple alerts. **Max alerts we will show at once is 3.**

![1 2 5 Multiple Alerts](https://user-images.githubusercontent.com/8203939/89795687-ae577a80-daf6-11ea-874d-4112d993c98e.png)

# Goal 2 & 3 (changes to nav)

- ### [Goal 2: Show save status clearly in nav](https://github.com/LLK/scratch-gui/issues/6080)
- ### [Goal 3: Show connectivity more clearly](https://github.com/LLK/scratch-gui/issues/6081)

---------
# What is in Phase II?
At this time, we have decided to readdress the issues below in Phase II. Not yet scheduled.

### Resolve File menu action behaviors
- Right now, you can click "Save as Copy," "New," "Load Project" etc in the File menu, but it will not let you leave your project session if you have a save error
- We should design a way to let users perform these actions, or disable them in the menu, depending on what makes sense per context

### Asset too big: Preventative alerts

- Should we tell the user at the time of upload if the asset is too big?
- Should we prevent an upload?

### Revisit decision around sound editor buttons

- Right now you can't edit a sound if the edit will make it too big.
- Is this the behavior we want?
- If no, we should redesign.
- If yes, we should make sure we tell the user what is happening. Currently no feedback.

### Revisit nav state when your account session changes
- Currently, when you get logged out in another tab, we still show you signed in
- We should communicate to people when
   - They get logged out (no feedback right now)
   - They get logged out and someone else is logged in (no feedback right now, but you [do get this error](https://github.com/LLK/scratch-gui/issues/5980) if you are trying to save someone else's project, which is not ideal)

### Revisit "Save now" in nav
- [More details here](https://github.com/LLK/scratch-gui/issues/6080)
- The styling of the "Save now" action as a text component is inconsistent with other actionable buttons.
- We would like to remove, restyle, or rethink this button eventually
- For now, we will not change this behaviorally or visually, because Scratchers still rely on this UI to save. Thanks to @Za-Chary on Scratch for raising this

**Comment by apple502j** [comment]
Technical stuff from a person who tried to implement it ( #5768 )

It's important to know that one asset can have multiple names. If you just have `Asset` then you need to [find the name of the asset by looping](https://github.com/LLK/scratch-gui/pull/5768/files#diff-4f8617ebfc2aacdb7b59f4ab230f473dR128), which may be a performance concern.

Also, `lib/alerts/index.jsx` expects the message to be static. If it needs to contain asset name, messages need to be generated dynamically. I did it by accepting a function as an argument and having `alert.jsx` handle it.

Finally the big problem is: there are many ways this can fail. It can be a number from FetchTool (e.g. 403 (invalid session), 413 (too big), 500 (server error)), it can be `Error` subclass (`JSON.parse` or `NetworkError`), or maybe string (although I think it's just NetsTool)

These are tricky parts but I'm sure ST can do it.

**Comment by Explosion-Scratch** [comment]
Wow! This looks great! The mockups are really great and the new styles also look helpfull. I also like being able to exit the alert.

**Comment by towerofnix** [comment]
One design comment:

> Max alerts we will show at once is 3.

Perhaps a small "..." ellipsis bubble could be shown beneath the third reminder, if there is more than three present at once?

**Comment by Explosion-Scratch** [comment]
Also this may be unrelated, but why is there "see community" instead of "see project page"?

**Comment by towerofnix** [comment]
@Explosion-Scratch It's a slightly different design from the actual implementation (out of date probably); you can spot more slight inconsistencies in places like the costume editor toolbox. But yeah, those don't have anything to do with the slide notes regarding the save alerts project.

**Comment by mxmou** [comment]
> @Explosion-Scratch It's a slightly different design from the actual implementation (out of date probably); you can spot more slight inconsistencies in places like the costume editor toolbox. But yeah, those don't have anything to do with the slide notes regarding the save alerts project.

The bitmap/vector switch and big block category names actually seem better than what we currently have.

**Comment by Explosion-Scratch** [comment]
@Mxmou, Probably lol.

**Comment by JeffaloBob** [comment]
those are displayed on smaller displays

**Comment by kaaaaathy** [comment]
> @Explosion-Scratch It's a slightly different design from the actual implementation (out of date probably); you can spot more slight inconsistencies in places like the costume editor toolbox. But yeah, those don't have anything to do with the slide notes regarding the save alerts project.

Yeah good observation :) It is a bit inconsistent based on what files I was using. You can ignore anything that doesn't look like what's in production right now

**Comment by griffpatch** [comment]
Is it worth considering adding a confirmation on delete of sprite, and
sharing of your project while you are looking into these?

**Comment by Explosion-Scratch** [comment]
@griffpatch Probably but those might be for another issue, however, I would love those features as well.

**Comment by mxmou** [comment]
Any progress on this?

**Comment by mxmou** [comment]
@chrisgarrity ^^

**Comment by the2000isawesome** [comment]
Has any progress been made towards fixing this problem since the last posts?

**Comment by the2000isawesome** [comment]
Any updates?

**Comment by leahcimto** [comment]
Any progress on this?

**Comment by hiPeeps124816** [comment]
this shouldn't be too hard to implement, any more progress yet?

**Comment by apple502j** [comment]
Due to how saving works (and how it is done on three different components - WWW, GUI and Storage) this is actually more complicated than you imagine.

But, this is not impossible; should definitely be implemented.
